### PR TITLE
Extract Scale link Bias link from caffe_function

### PR DESCRIPTION
**Pull request by takagi**
Rel #1239.

### CODE REVIEW

**Comment by takagi** [comment]
Status ::= Done | Doing | Open

|Item|Documentation|Test|Implementation|
|---|---|---|---|
|`scale` function|Open|Open|Open|
|`Scale` chain|Open|Open|Open|
|`bias` function|Open|Open|Open|
|`Bias` link|Open|Open|Open|

**Comment by takagi** [comment]
Please review this PR, AppVeyor will pass soon as well as Travis.

File: chainer/functions/math/bias.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    """Elementwise summation with broadcasting.
+
+    Computes a elementwise summation of two input variables, with the shape of
+    the latter variable broadcasted to match the shape of the former. `axis`
+    is the first axis of the first variable along which to apply the second
+    variable.
+
+    Args:
+        x (~chainer.Variable): Input variable to be scaled.
+        y (~chainer.Variable): Input variable to scale, broadcasted.
```
Comment (delta2323) [inline chainer/functions/math/bias.py]: I think the description of `x` and `y` is not appropriate as arguments of `bias` function.

File: chainer/functions/math/bias.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+        axis (int): The first axis of `x` along which to apply `y`
+
+    Returns:
+        ~chainer.Variable: Output variable.
+
+    """
+    x_shape = x.data.shape
+    y_shape = y.data.shape
+    assert x_shape[axis:axis + len(y_shape)] == y_shape
```
Comment (delta2323) [inline chainer/functions/math/bias.py]: Should enable this assertion only when debug mode? We can switch the mode with `chainer.set_debug` method.

File: chainer/functions/math/bias.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+    Computes a elementwise summation of two input variables, with the shape of
+    the latter variable broadcasted to match the shape of the former. `axis`
+    is the first axis of the first variable along which to apply the second
+    variable.
+
+    Args:
+        x (~chainer.Variable): Input variable to be scaled.
+        y (~chainer.Variable): Input variable to scale, broadcasted.
+        axis (int): The first axis of `x` along which to apply `y`
```
Comment (delta2323) [inline chainer/functions/math/bias.py]: I think "The first axis of `x` which `y` is applied along" is more grammatically appropriate.

File: chainer/functions/math/scale.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+    Computes a elementwise product of two input variables, with the shape of
+    the latter variable broadcasted to match the shape of the former. `axis`
+    is the first axis of the first variable along which to apply the second
+    variable.
+
+    Args:
+        x (~chainer.Variable): Input variable to be scaled.
+        y (~chainer.Variable): Input variable to scale, broadcasted.
+        axis (int): The first axis of `x` along which to apply `y`
```
Comment (delta2323) [inline chainer/functions/math/scale.py]: ditto ( "The first axis of x along which y is applied".)

File: chainer/links/caffe/caffe_function.py [hunk 5]
```diff
@@ -371,14 +371,14 @@ def _setup_scale(self, layer):
         # Case of only one bottom where W is learnt parameter.
         if len(bottom) == 1:
             W_shape = blobs[0].shape.dim
-            func = _Scale(axis, W_shape, bias_term)
+            func = links.Scale(axis, W_shape, bias_term)
```
Comment (delta2323) [inline chainer/links/caffe/caffe_function.py]: Please do not use alias defined in `chainer.links.__init__` inside `Chainer` module. Use `links.scale.Scale` instead.

File: chainer/links/caffe/caffe_function.py [hunk 6]
```diff
@@ -371,14 +371,14 @@ def _setup_scale(self, layer):
...
             func.W.data.ravel()[:] = blobs[0].data
             if bias_term:
                 func.bias.b.data.ravel()[:] = blobs[1].data
         # Case of two bottoms where W is given as a bottom.
         else:
             shape = blobs[0].shape.dim if bias_term else None
-            func = _Scale(axis, bias_term=bias_term, bias_shape=shape)
+            func = links.Scale(axis, bias_term=bias_term, bias_shape=shape)
```
Comment (delta2323) [inline chainer/links/caffe/caffe_function.py]: ditto (use `links.scale.Scale`)

File: chainer/links/connection/bias.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
+from chainer.functions.math import bias
+from chainer import link
+
+
+class Bias(link.Link):
+    """Broadcasted elementwise summation with learnable parameter.
+
+    Computes a elementwise summation as :func:`~chainer.functions.bias`
+    function except that its second input is a learnable bias parameter
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: as ~chainer.functions.bias function *does*

File: chainer/links/connection/bias.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    :math:`b` the link has.
+
+    Args:
+        axis (int): The first axis of the first input of
+            :func:`~chainer.functions.bias` function along which to apply its
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: ditto

File: chainer/links/connection/bias.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+            self.b.data.fill(0)
+
+        # Hold axis.
+        self.axis = axis
+
+    def __call__(self, *xs):
+        """Applies broadcasted elementwise summation.
+
+        Args:
+            *xs (list of ~chainer.Variable): Input variables whose length
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: * is not needed.

File: chainer/links/connection/scale.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+class Scale(link.Chain):
+    """Broadcasted elementwise product with learnable parameter.
+
+    Computes a elementwise product as :func:`~chainer.functions.scale`
+    function except that its second input is a learnable weight parameter
+    :math:`W` the link has.
+
+    Args:
+        axis (int): The first axis of the first input of
+            :func:`~chainer.functions.scale` function along which to apply its
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto

File: chainer/links/connection/bias.py [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    def __init__(self, axis=1, shape=None):
+        super(Bias, self).__init__()
+
+        # Add b parameter if given.
+        if shape is not None:
+            self.add_param('b', shape)
+            self.b.data.fill(0)
+
+        # Hold axis.
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: I think this comment is not needed as we can understand it from the following line.

File: chainer/links/connection/scale.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+                self.add_link('bias', func)
+        else:
+            if bias_term:
+                if bias_shape is None:
+                    raise ValueError('bias_shape should be given if W is not '
+                                     'learnt parameter and bias_term is True.')
+                func = bias.Bias(axis, bias_shape)
+                self.add_link('bias', func)
+
+        # Hold axis.
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto (the comment is not needed.)

File: chainer/links/connection/scale.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    def __init__(self, axis=1, W_shape=None, bias_term=False, bias_shape=None):
+        super(Scale, self).__init__()
+
+        # Add W parameter if given.
+        if W_shape is not None:
+            self.add_param('W', W_shape)
+            self.W.data.fill(1)
+
+        # Add bias term if given.
+        if W_shape is not None:
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: These two if statements can be merged.

File: chainer/links/connection/scale.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+            if bias_term:
+                func = bias.Bias(axis, W_shape)
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: `W_shape` in this line should be `bias_shape`?

File: chainer/links/connection/scale.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+                self.add_link('bias', func)
+
+        # Hold axis.
+        self.axis = axis
+
+    def __call__(self, *xs):
+        """Applies broadcasted elementwise product.
+
+        Args:
+            *xs (list of ~chainer.Variable): Input variables whose length
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto (remove asterisk)

File: chainer/links/connection/bias.py [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+        Args:
+            xs (list of ~chainer.Variable): Input variables whose length should
+                be one if the link has a learnable bias parameter, otherwise
+                should be two.
+        """
+        axis = self.axis
+
+        # Case of only one bottom where b is learnt parameter.
+        if hasattr(self, 'b'):
+            assert len(xs) == 1
```
Comment (takagi) [inline chainer/links/connection/bias.py]: Also better to assert only in debug mode?

File: chainer/links/connection/bias.py [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+            x, = xs
+            b = self.b
+            return bias.bias(x, b, axis)
+        # Case of two bottoms where b is given as a bottom.
+        else:
+            assert len(xs) == 2
```
Comment (takagi) [inline chainer/links/connection/bias.py]: Also better to assert only in debug mode?

File: chainer/links/connection/scale.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        Args:
+            xs (list of ~chainer.Variable): Input variables whose length should
+                be one if the link has a learnable weight parameter, otherwise
+                should be two.
+        """
+        axis = self.axis
+
+        # Case of only one bottom where W is learnt parameter.
+        if hasattr(self, 'W'):
+            assert len(xs) == 1
```
Comment (takagi) [inline chainer/links/connection/scale.py]: Also better to assert only in debug mode?

File: chainer/links/connection/scale.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+            x, = xs
+            W = self.W
+            z = scale.scale(x, W, axis)
+        # Case of two bottoms where W is given as a bottom.
+        else:
+            assert len(xs) == 2
```
Comment (takagi) [inline chainer/links/connection/scale.py]: Also better to assert only in debug mode?

**Comment by delta2323** [comment]
Thank you for your comment. I think it is enough that assertion on the length of `xs` should be done only in debug mode.

**Comment by takagi** [comment]
Okay. I will fix them shortly.

File: chainer/__init__.py [hunk 20]
```diff
@@ -76,6 +76,28 @@ def set_debug(debug):
     global _debug
     _debug = debug
 
+
+class DebugMode(object):
+    """Debug mode context.
+
+    This class provides a context manager for debug mode. When entering the
+    context, it sets the debug mode to the value of `debug` parameter with
+    memorizing its original value. When exiting the context, it set the debug
```
Comment (delta2323) [inline chainer/__init__.py]: it set*s*

**Comment by delta2323** [comment]
We should note that broadcasting in the context of `Bias` and `Scale` is different from one used in other codes of Chainer, as you wrote in the [comment](https://github.com/pfnet/chainer/pull/1028#issuecomment-205260767) of PR #1028. We should add this note to somewhere in the comments of `Bias` and `Scale`.

File: tests/chainer_tests/links_tests/connection_tests/test_bias.py [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+from chainer import links
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+@testing.parameterize(*[
+    {'learn_b': True},
+    {'learn_b': False},
+])
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_bias.py]: The following code should works

```
@testing.parameterize(
   {'learn_b': True},
   {'learn_b': False}
)
```

File: tests/chainer_tests/links_tests/connection_tests/test_bias.py [hunk 22]
```diff
@@ -0,0 +1,119 @@
...
+
+    def setUp(self):
+        self.x = numpy.random.uniform(-1, 1, (3, 2, 3)).astype(numpy.float32)
+        self.b = numpy.random.uniform(-1, 1, (2)).astype(numpy.float32)
+        self.y_expected = numpy.copy(self.x)
+        for i, j, k in numpy.ndindex(self.y_expected.shape):
+            self.y_expected[i, j, k] += self.b[j]
+        self.gy = numpy.random.uniform(-1, 1, (3, 2, 3)).astype(numpy.float32)
+
+        learn_b = self.learn_b
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_bias.py]: This assignment is redundant.

File: tests/chainer_tests/links_tests/connection_tests/test_bias.py [hunk 23]
```diff
@@ -0,0 +1,119 @@
...
+        axis = 1
+        if learn_b:
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_bias.py]: Use `self.learn_b`

File: tests/chainer_tests/links_tests/connection_tests/test_scale.py [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+@testing.parameterize(*[
+    {'learn_W': True, 'bias_term': False, 'bias_shape': None},
+    {'learn_W': True, 'bias_term': True, 'bias_shape': None},
+    {'learn_W': False, 'bias_term': False, 'bias_shape': None},
+    {'learn_W': False, 'bias_term': True, 'bias_shape': (2,)},
+])
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_scale.py]: ditto (remove `*[` and `]`)

File: tests/chainer_tests/links_tests/connection_tests/test_scale.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+        self.W = numpy.random.uniform(-1, 1, (2)).astype(numpy.float32)
+        self.b = numpy.random.uniform(-1, 1, (2)).astype(numpy.float32)
+        self.y_expected = numpy.copy(self.x)
+        for i, j, k in numpy.ndindex(self.y_expected.shape):
+            self.y_expected[i, j, k] *= self.W[j]
+            if self.bias_term:
+                self.y_expected[i, j, k] += self.b[j]
+        self.gy = numpy.random.uniform(-1, 1, (3, 2, 3)).astype(numpy.float32)
+
+        learn_W = self.learn_W
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_scale.py]: ditto (this assignment is redundant).

File: tests/chainer_tests/links_tests/connection_tests/test_scale.py [hunk 26]
```diff
@@ -0,0 +1,141 @@
...
+        bias_term = self.bias_term
+        bias_shape = self.bias_shape
+        axis = 1
+        if learn_W:
```
Comment (delta2323) [inline tests/chainer_tests/links_tests/connection_tests/test_scale.py]: use `self.learn_W`

**Comment by delta2323** [comment]
LGTM except comments.

**Comment by delta2323** [comment]
Please merge or rebase the master branch after applying the comments.

**Comment by takagi** [comment]
> We should note in somewhere in the comments of Bias and Scale that broadcasting in the context of these classes is different from one used in other codes of Chainer, as you wrote in the comment of PR #1028.

I noted it in scale function's and bias function's docstring.

**Comment by takagi** [comment]
I've merged the master.

File: chainer/links/connection/bias.py [hunk 27]
```diff
@@ -0,0 +1,59 @@
+import chainer
+from chainer.functions.math import bias
+from chainer import link
+
+
+class Bias(link.Link):
+    """Broadcasted elementwise summation with learnable parameter.
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: parameter*s*

File: chainer/links/connection/bias.py [hunk 28]
```diff
@@ -0,0 +1,59 @@
...
+        """Applies broadcasted elementwise summation.
+
+        Args:
+            xs (list of ~chainer.Variable): Input variables whose length should
+                be one if the link has a learnable bias parameter, otherwise
+                should be two.
+        """
+        axis = self.axis
+
+        # Case of only one bottom where b is learnt parameter.
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: "bottom" is a Caffe term. So could you use "argument" instead?

File: chainer/links/connection/bias.py [hunk 29]
```diff
@@ -0,0 +1,59 @@
...
+        if hasattr(self, 'b'):
+            if chainer.is_debug():
+                assert len(xs) == 1
+            x, = xs
+            b = self.b
+            return bias.bias(x, b, axis)
+        # Case of two bottoms where b is given as a bottom.
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: ditto (use `arguments` instead of bottoms)

File: chainer/links/connection/scale.py [hunk 30]
```diff
@@ -0,0 +1,83 @@
...
+        """Applies broadcasted elementwise product.
+
+        Args:
+            xs (list of ~chainer.Variable): Input variables whose length should
+                be one if the link has a learnable weight parameter, otherwise
+                should be two.
+        """
+        axis = self.axis
+
+        # Case of only one bottom where W is learnt parameter.
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto (use `argument` instead of bottom)

File: chainer/links/connection/scale.py [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+        if hasattr(self, 'W'):
+            if chainer.is_debug():
+                assert len(xs) == 1
+            x, = xs
+            W = self.W
+            z = scale.scale(x, W, axis)
+        # Case of two bottoms where W is given as a bottom.
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto (use `arguments` instead of bottoms)

File: chainer/links/connection/bias.py [hunk 32]
```diff
@@ -0,0 +1,59 @@
...
+    :math:`b` the link has.
+
+    Args:
+        axis (int): The first axis of the first input of
+            :func:`~chainer.functions.bias` function along which its second
+            input is applied.
+        shape (tuple of ints): Shape of the learnable bias parameter. If
+            ``None``, a bias parameter needs to be given explicitly to its
+            ``__call__`` method's second input.
+
```
Comment (delta2323) [inline chainer/links/connection/bias.py]: We should note that this link does not have parameters to be learnt if `shape` is `None`.

File: chainer/links/connection/scale.py [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+    function does except that its second input is a learnable weight parameter
+    :math:`W` the link has.
+
+    Args:
+        axis (int): The first axis of the first input of
+            :func:`~chainer.functions.scale` function along which its second
+            input is applied.
+        W_shape (tuple of ints): Shape of learnable weight parameter. If
+            ``None``, a weight parameter needs to be given explicitly to its
+            ``__call__`` method's second input.
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: We should note that `W` is not a learnt parameter if `W_shape` is `None`.

File: chainer/links/connection/scale.py [hunk 34]
```diff
@@ -0,0 +1,83 @@
...
+        bias_shape (tuple of ints): Shape of learnable bias. If ``W_shape`` is
+            ``None``, this should be given to determine the shape. Otherwise,
+            the bias has the same shape ``W_shape`` with the weight parameter
+            and ``bias_shape`` is ignored.
+
+    .. seealso:: See :func:`chainer.functions.scale` for details.
+
+    Attributes:
+        W (~chainer.Variable): Weight parameter.
+        bias (~chainer.links.Bias): Bias term.
```
Comment (delta2323) [inline chainer/links/connection/scale.py]: ditto (`W` and `bias` parameters can not exist depending on the init parameters).

**Comment by delta2323** [comment]
LGTM 👍

**Pull request merged**